### ISSUE TITLE
restic with "cold" storage (here: OVH Cloud Archive)

### ISSUE THREAD

**Issue by aawsome** [restic/restic#2504]
Output of `restic version`
--------------------------
restic 0.9.6 compiled with go1.13.4 on linux/amd64

How did you run restic exactly?
-------------------------------
`restic -r rclone:ovh:test2 -p key-restic snapshots`

Output is:
```
Load(<key/1bd3e4429a>, 0, 0) returned error, retrying after 720.254544ms: <key/1bd3e4429a> does not exist
```
repeated a couple of times with changing retry times and then:
`Fatal: <key/1bd3e4429a> does not exist`

When using
`RCLONE_VERBOSE=2 restic -r rclone:ovh:test2 -p key-restic snapshots`
you see:
```
rclone: 2019/12/07 12:40:04 DEBUG : Swift container test2: GET /file-5577006791947779410
rclone: 2019/12/07 12:40:04 DEBUG : file-5577006791947779410: GET request error: object not found
rclone: 2019/12/07 12:40:04 DEBUG : Swift container test2: HEAD /config
rclone: 2019/12/07 12:40:05 DEBUG : Swift container test2: GET /keys/
rclone: 2019/12/07 12:40:05 DEBUG : keys: list request
rclone: 2019/12/07 12:40:06 DEBUG : Swift container test2: GET /keys/1bd3e4429a441fd7546efdfa1bc7c41333d548b0a1ecb5dea3fd1f9ef16c205f
rclone: 2019/12/07 12:40:07 DEBUG : 1bd3e4429a441fd7546efdfa1bc7c41333d548b0a1ecb5dea3fd1f9ef16c205f: Get request open error: try again after 2019-12-07T12:50:07.317400358+01:00 (9m59.999921857s)
Load(<key/1bd3e4429a>, 0, 0) returned error, retrying after 720.254544ms: <key/1bd3e4429a> does not exist
```

Similar behavior with other restic commands, like `backup` etc.

If you wait long enough and redo the command, restic works fine. However, the issue occurs with

- key files
- config file
- lock files

Therefore you might have to repeat and wait quite often before the requested operation is performed.

Another issue is, that the above files are not cached and therefore once you manage to get these files "warm" in the backend, you can do your operations. But if they get "cold" again, you again come in the original mentioned issue. 

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
rclone to access OVH via swift.
The rclone config is directly as provided by OVH.
test2 is a bucket in [OVH Cloud Archive](https://www.ovh.com/world/public-cloud/cloud-archive/)

Expected behavior
-----------------

- restic should wait until the data is ready to access. Ideally giving a hint that it is waiting.
- restic should cache config file, key files and lock files.

Actual behavior
---------------

- restic stops without performing the requested observation when data is not yet ready to access
- config file, key files and lock files are not cached.

Steps to reproduce the behavior
-------------------------------

- create an bucket in OVH Cloud Archive
- run restic init on this bucket
- wait some time to let the data become "cold"
- perform the commands given above

Do you have any idea what may have caused this?
-----------------------------------------------
restic is not prepared to using a "cold" backend.

Do you have an idea how to solve the issue?
-------------------------------------------
1. I would suggest adding an option which allows restic to wait even for a real long time if it doesn't receive the data for a backend
2. Even better would be to get and use the information how long we have to wait. Then just wait long enough and display to the user that you are waiting.
3. Another improvement would be to use the waiting time an try to get other files you need from the backend. Ideally restic should try to start the warmup-process for as many files as possible.
4. Implement caching for key files, the config file and lock files. I think this involves following steps:

- Optionally organize caches by repo string instead of repo ID to make caching of key files and config files possible
- Add key files, config files and lock files to the cache implementation in `internal/cache`
- Check which commands need improvement to use the cache information

I would like to work on 1. and 4. and appreciate any help!

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
I do my daily backups with restic and like it very much! I especially like the possibility to directly backup to a cloud storage. Before I used other backup tools to backup to a local disc and sync the backup to a cloud storage. This solution failed badly once I had a problem with my USB hub which destroyed my local disc backup (and consequently my cloud storage copy)
Moreover using restic backup --stdin for DB backups is just great and makes me very happy :smile:

**Comment by aawsome** [comment]
To summarize what I have done:

- PR #2515 adds a `--retries` flag. Using `--retries -1`solves the first point of this issue.
- PR #2516 adds a '--cache-all' flag. Together with PR #2505 also the config file is cached. Using `--cache-all` then solves the 4th point.
- Additionally, PR #2513 adds the commands 'cleanup-index' and 'cleanup-packs'. With cold storage, you do not want to repack the pack files (needs to read the pack files). So 'cleanup-index' followed by 'cleanup-packs' serve as a replacement for 'prune' with cold storages.
- If you are in an emergency and have packs missing in the index, PR #2507 optimizes 'rebuild-index' so that only the packs not in the index needs to be read.

As these are quite some changes, I made a new branch ["cold-storage"](https://github.com/aawsome/restic/tree/cold-storage) which includes all the above mentioned PRs. This works like a charm with OVH Cloud Archive :smile: 

I suggest to close this issue when PRs  #2505, #2515, #2516 are merged.

**Comment by rawtaz** [comment]
Has this and the changes you made PRs for been discussed anywhere before?

**Comment by aawsome** [comment]
I did open this issue 9 days ago hoping for some discussion. As I had some free time and enjoy hacking, I tried to find the relevant pieces of code and started programming. Open source rocks :+1: 

Please write if you have any comments or other/better ideas. I'm looking forward to improve my PRs if I can contribute to a better restic!

**Comment by rawtaz** [comment]
These things need discussion. We can't just go adding flags and commands without very careful consideration, because if we do so we'll end up with a complete mess. In this case you've suggested adding *four* new flags and commands in total.

Now, there's of course nothing wrong with writing code if you like that, but be prepared that the PRs won't just get merged out of the blue :)

Thanks for the summary above, that's a good starting point. I'm labeling this as discussion for now.

**Comment by aawsome** [comment]
Thank you for your feedback!

Of course the naming of commands and flags can and should be discussed. I care much more about the functionality than about the naming and I'm happy to replace the naming if there is something more reasonable. I'm also happy to get feedback about the code and look forward to dealing with issues about my proposals.

The reason why I made two new commands was that I had the feeling there are quite some issues with `prune` and that somehow `prune` should be re-engineered. I started implementation as it shows that this also helps with the "cold storage" handling. But maybe a discussion in a separate issue would be more fitting... However, there has been quite some issues about prune (some of which are quite old) I didn't find a good issue to discuss and thought I'd just give it a shot to start with a proposal of an implementation.

I split the work in many PRs as they can all be independently merged and - to my feeling - most improve restic by itself. So if only some are merged and others will be regarded, there will still be some improvement to restic. Also I had the feeling that separating code changes in independent PRs makes reviewing easier...

**Comment by aawsome** [comment]
Here's an update to the points mentioned above:

- #2515 is still open, but changed such that you have to specify a positive number of retries
- I made #2516 a draft. I'm working on an option to use a arbitrary repository (may be a remote but not cold storage) as cache. This will also solve the issues with #2516 that it's only local and that the cache dir name might contain bad information. However, as a workaround, #2516 together with #2505 still works.
- #2513 is superseded by #2718

**Comment by bfontaine** [comment]
Should this be closed in favor of #3202?

**Comment by aawsome** [comment]
Yes, closing this as it is superseded by #3202
